File: src/main/java/gg/mineral/practice/tournaments/Tournament.java
package gg.mineral.practice.tournaments;

import gg.mineral.api.collection.GlueList;
import gg.mineral.practice.PracticePlugin;
import gg.mineral.practice.bukkit.events.PlayerTournamentInitializeEvent;
import gg.mineral.practice.bukkit.events.PlayerTournamentStartEvent;
import gg.mineral.practice.entity.Profile;
import gg.mineral.practice.managers.ProfileManager;
import gg.mineral.practice.managers.TournamentManager;
import gg.mineral.practice.match.Match;
import gg.mineral.practice.match.TournamentMatch;
import gg.mineral.practice.match.data.MatchData;
import gg.mineral.practice.util.collection.ProfileList;
import gg.mineral.practice.util.messages.impl.ChatMessages;
import gg.mineral.practice.util.messages.impl.ErrorMessages;
import lombok.Getter;
import lombok.val;
import net.md_5.bungee.api.chat.ClickEvent;
import org.bukkit.Bukkit;
import org.bukkit.scheduler.BukkitRunnable;

public class Tournament {
    GlueList<Match> matches = new GlueList<>();
    ProfileList players = new ProfileList();
    boolean started = false;
    @Getter
    boolean ended = false;
    MatchData matchData;
    int round = 1;
    @Getter
    String host;

    public Tournament(Profile p) {
        this.matchData = new MatchData(p.getDuelSettings());
        this.host = p.getName();
        addPlayer(p);
    }

    public void addPlayer(Profile p) {

        if (started) {
            p.message(ErrorMessages.TOURNAMENT_STARTED);
            return;
        }

        p.setTournament(this);
        players.add(p);

        val joinedMessage = ChatMessages.JOINED_TOURNAMENT.clone().replace("%player%", p.getName());
        ProfileManager.broadcast(players, joinedMessage);
    }

    public void removePlayer(Profile p) {
        players.remove(p);
        p.setTournament(null);

        val leftMessage = ChatMessages.LEFT_TOURNAMENT.clone().replace("%player%", p.getName());
        ProfileManager.broadcast(players, leftMessage);

        if (players.isEmpty()) {
            TournamentManager.remove(this);
            ended = true;
            return;
        }

        if (started && players.size() == 1) {
            val winner = players.getFirst();
            winner.removeFromTournament();

            TournamentManager.remove(this);
            ended = true;

            val wonMessage = ChatMessages.WON_TOURNAMENT.clone().replace("%player%", winner.getName());

            ProfileManager.broadcast(wonMessage);

        }
    }

    public void start() {

        if (started)
            return;

        val bukkitHost = players.getFirst().getPlayer();

        val event = new PlayerTournamentInitializeEvent(30, bukkitHost);
        Bukkit.getPluginManager().callEvent(event);

        if (event.isCancelled())
            return;

        TournamentManager.registerTournament(this);

        val messageToBroadcast = ChatMessages.BROADCAST_TOURNAMENT.clone()
                .replace("%player%", host).setTextEvent(new ClickEvent(ClickEvent.Action.RUN_COMMAND, "/join " + host),
                        ChatMessages.CLICK_TO_JOIN);

        ProfileManager.broadcast(messageToBroadcast);

        new BukkitRunnable() {
            @Override
            public void run() {
                started = true;

                val event = new PlayerTournamentStartEvent(bukkitHost);
                Bukkit.getPluginManager().callEvent(event);

                if (players.size() == 1 || event.isCancelled()) {
                    Profile winner = players.getFirst();
                    winner.removeFromTournament();

                    ErrorMessages.TOURNAMENT_NOT_ENOUGH_PLAYERS.send(winner.getPlayer());
                    TournamentManager.remove(Tournament.this);
                    ended = true;
                    return;
                }

                startRound();
            }
        }.runTaskLater(PracticePlugin.INSTANCE, 600);
    }

    public void startRound() {

        if (players.size() == 1) {
            Profile winner = players.getFirst();
            winner.removeFromTournament();
            ended = true;
            TournamentManager.remove(this);
            return;
        }

        val iter = players.iterator();

        while (iter.hasNext()) {
            val p1 = iter.next();

            if (!iter.hasNext()) {
                ChatMessages.NO_OPPONENT.send(p1.getPlayer());
                return;
            }

            val p2 = iter.next();

            val match = new TournamentMatch(p1, p2, matchData, this);
            match.start();
            matches.add(match);
        }
    }

    public void removeMatch(Match m) {
        matches.remove(m);

        if (ended)
            return;

        if (matches.isEmpty()) {
            val broadcastedMessage = ChatMessages.ROUND_OVER.clone().replace("%round%", "" + round);

            ProfileManager.broadcast(players, broadcastedMessage);

            new BukkitRunnable() {
                @Override
                public void run() {
                    startRound();
                    round++;
                }
            }.runTaskLater(PracticePlugin.INSTANCE, 100);
        }
    }
}

File: src/main/kotlin/gg/mineral/practice/tournaments/Tournament.kt
package gg.mineral.practice.tournaments

import gg.mineral.practice.contest.Contest
import gg.mineral.practice.duel.DuelSettings
import gg.mineral.practice.entity.Profile
import gg.mineral.practice.managers.ContestManager.registerContest
import gg.mineral.practice.managers.ContestManager.remove
import gg.mineral.practice.managers.GametypeManager
import gg.mineral.practice.managers.ProfileManager.broadcast
import gg.mineral.practice.match.TournamentMatch
import gg.mineral.practice.match.data.MatchData
import gg.mineral.practice.util.math.Countdown
import gg.mineral.practice.util.messages.impl.ChatMessages
import gg.mineral.practice.util.messages.impl.ErrorMessages
import net.md_5.bungee.api.chat.ClickEvent
import org.bukkit.Bukkit

open class Tournament(
    private val hostName: String,
    open val waitTime: Int = 60,
    override val maxPlayers: Int = 24,
    override val matchData: MatchData = MatchData(DuelSettings(null, GametypeManager.gametypes.values.randomOrNull()))
) :
    Contest(hostName) {

    override fun addPlayer(p: Profile): Boolean {
        if (participants.size >= maxPlayers) return p.message(ErrorMessages.TOURNAMENT_FULL).let { true }
        if (started) return p.message(ErrorMessages.TOURNAMENT_STARTED).let { true }
        p.contest = this
        return participants.add(p)
            .also { broadcast(participants, ChatMessages.JOINED_TOURNAMENT.clone().replace("%player%", p.name)) }
    }

    override fun removePlayer(p: Profile) {
        if (!participants.remove(p)) return
        broadcast(participants, ChatMessages.LEFT_TOURNAMENT.clone().replace("%player%", p.name))
        p.contest = null
        if (!started) return
        if (participants.isEmpty()) {
            onContestEnd()
            remove(this)
        } else if (participants.size == 1) {
            val winner = participants.first
            winner?.contest = null
            onContestWin(winner)
            broadcast(ChatMessages.WON_TOURNAMENT.clone().replace("%player%", winner!!.name))
        }
    }

    override fun startContest() {
        if (started) return
        if (registerContest(this) != null) return

        val countdown = Countdown(waitTime, this) { time ->
            if (time % 30 != 0) return@Countdown
            broadcast(ChatMessages.TOURNAMENT_BROADCAST)
            Bukkit.getPlayer(hostName)?.let {
                broadcast(
                    ChatMessages.CONTEST_HOST.clone().replace(
                        "%host%",
                        hostName
                    )
                )
            }

            matchData.gametype?.name?.let {
                broadcast(
                    ChatMessages.CONTEST_MODE.clone().replace(
                        "%mode%",
                        it
                    )
                )
            }

            broadcast(
                ChatMessages.CONTEST_PLAYERS.clone().replace(
                    "%players%",
                    participants.size.toString() + "/" + maxPlayers
                )
            )

            if (this is AutomatedTournament)
                broadcast(ChatMessages.CONTEST_REWARD.clone().replace("%rank%", reward.rank))

            fun formatSeconds(seconds: Int): String {
                val minutes = seconds / 60
                val remainingSeconds = seconds % 60
                return "${minutes}m ${remainingSeconds}s"
            }

            broadcast(
                ChatMessages.CONTEST_STARTS_IN.clone().replace(
                    "%time%",
                    formatSeconds(time)
                )
            )

            broadcast(
                ChatMessages.CONTEST_JOIN.clone().setTextEvent(
                    ClickEvent(ClickEvent.Action.RUN_COMMAND, "/join $hostName"),
                    ChatMessages.CLICK_TO_JOIN
                )
            )
        }
        countdown.start()
    }

    override fun createMatch(p1: Profile, p2: Profile) = TournamentMatch(p1, p2, matchData, this)

    override fun onContestWin(winner: Profile?) {
        winner?.contest = null
        ended = true
        remove(this)
    }

    override fun onContestEnd() {
        ended = true
        remove(this)
    }

    override fun onStart(profile: Profile) {
    }

    override fun onStart() {
        started = true
        if (participants.size == 1) {
            val winner = participants.first
            winner?.contest = null
            winner?.message(ErrorMessages.TOURNAMENT_NOT_ENOUGH_PLAYERS)
            onContestEnd()
        } else startRound()
    }

    override fun onCountdownStart(profile: Profile) {
    }
}
